rust/arrow/src/array/array.rs

```diff
@@ -2851,4 +3153,50 @@ mod tests {
         assert!(ret.is_ok());
         assert_eq!(8, ret.ok().unwrap());
     }
+
+    #[test]
+    fn test_dictionary_array_fmt_debug() {
+        let key_builder = PrimitiveBuilder::<UInt8Type>::new(3);
+        let value_builder = PrimitiveBuilder::<UInt32Type>::new(2);
+        let mut builder = PrimitiveDictionaryBuilder::new(key_builder, value_builder);
+        builder.append(12345678).unwrap();
+        builder.append_null().unwrap();
+        builder.append(22345678).unwrap();
+        let array = builder.finish();
+        assert_eq!(
+            "DictionaryArray {keys: [Some(0), None, Some(1)] values: PrimitiveArray<UInt32>\n[\n  12345678,\n  22345678,\n]}\n",
+            format!("{:?}", array)
+        );
+
+        let key_builder = PrimitiveBuilder::<UInt8Type>::new(20);
+        let value_builder = PrimitiveBuilder::<UInt32Type>::new(2);
+        let mut builder = PrimitiveDictionaryBuilder::new(key_builder, value_builder);
+        for _ in 0..20 {
+            builder.append(1).unwrap();
+        }
+        let array = builder.finish();
+        assert_eq!(
+            "DictionaryArray {keys: [Some(0), Some(0), Some(0), Some(0), Some(0), Some(0), Some(0), Some(0), Some(0), Some(0)]... values: PrimitiveArray<UInt32>\n[\n  1,\n]}\n",
+            format!("{:?}", array)
+        );
+    }
+
+    #[test]
+    fn test_dictionary_array_from_iter() {
+        let test = vec!["a", "a", "b", "c"];
+        let array: DictionaryArray<Int8Type> = test
+            .iter()
+            .map(|&x| if x == "b" { None } else { Some(x) })
+            .collect();
+        assert_eq!(
+            "DictionaryArray {keys: [Some(1), Some(1), Some(0), Some(2)] values: StringArray\n[\n  null,\n  \"a\",\n  \"c\",\n]}\n",
```

Comment:
I would have expected to see `"b"` as `None` in the `keys` and omit `"null"` from the `values`?

Reply:
There are two ways of doing nulls in dictionaries. One is to have nulls in the keys, but this adds 1 bit per key and a lot more logic. The other is to use one of the key values as a null - 0 is a good choice. So Some(0) represents a null value and it would be easier for SIMD code to use this. I would love to put in a values iterator for all array types and then we could iterate over list, struct and dict types.

Something like:
```
   let fred : DictionaryArray<Int8Type> = (0..19).map(|x| x % 10).collect();
   let fred2 : Vec<i32> = fred.iter().collect();
   assert_eq!(fred2, vec![0, 1, 2, ...9, 0, 1, 2... 9]);
```

Reply:
Right, you *could* represent it that way for sure but that's not following the Arrow [spec](https://arrow.apache.org/docs/format/Columnar.html#dictionary-encoded-layout), right?

What you are talking about above is the sentinel approach (where 0 is your sentinel), which has it's merits but Arrow choose the bitmap approach for better or worse.

IMHO iterator support and SIMD are secondary concerns with following the spec being more important.  Not sure how others feel but I would discuss this on the mailing list to gain some consensus first?

Reply:
Hi @paddyhoran @andy-thomason, I would also prefer that we conform to the spec on how we treat nulls, by using the bitmap approach.
I agree on discussing iterator support on the mailing list, my stance being that I don't mind a good API being introduced, but I'd prefer that being introduced in a separate PR.

_I'm preferring a separate PR_ because the Rust implementation currently doesn't have a lot of contributors, and so if we don't introduce functionality that covers all data types, I worry that:

* we might take a while before we get to cover other types, leaving an incomplete experience for users
* an API exposed on few types might need changing later if we find that it's not suitable for other types, leading to breakages.

This is partly why I've been working on large PRs that try to cover everything related to a feature (e.g. trying to support reading *everything we can* on IPC readers, trying to include new array types in compute).

Reply:
I'll remove the alternative encoding and the Uint* primitive types to make it conform to the standard.

Reply:
Thanks @andy-thomason.  Frustrating to put in the work and then removed it, appreciate your understanding.  Like I said if you feel strongly about it I would engage the wider community on the mailing list about it.

Reply:
No worries, Paddy. Just keen to get this thing in and stop having to merge it!

Reply:
Must say a big thanks to @nevi-me for all his hard work on the other PRs.

Reply:
@andy-thomason have you seen [this](https://lists.apache.org/thread.html/r17dab8c382fb30975568d416526215a50a635509d65a738e61020a44%40%3Cdev.arrow.apache.org%3E) discussion?  You might have some input on this as it relates to the bitmap vs sentinel approach for dictionary encoding.

Reply:
Thanks @paddyhoran, no time at the present for mailing lists, I'm afraid, but indeed it is relevant.